### ISSUE TITLE
[java] UnnecessaryModifier wrong message public->static

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#1356]
PMD version 6.7.0.

```java
public final class A {
	@FunctionalInterface
	public static interface B extends C<InputStream, IOException> {
	}

	...
}
```

This gives an `UnnecessaryModifier` which is wrong.

**Comment by adangel** [comment]
The reported message looks also wrong:

`n/a:3:	Unnecessary modifier 'public' on interface 'B': types nested within an interface type are implicitly static`

It should read "Unneccesary modifier **'static'** on interface 'B'..."

Apart from that, static is indeed not needed here: See https://docs.oracle.com/javase/specs/jls/se10/html/jls-8.html#jls-8.5.1

> A member interface is implicitly static (§9.1.1). It is permitted for the declaration of a member interface to redundantly specify the static modifier. 

So, I'd say, the issue is about the wrong message and it is not a false-positive.

**Comment by boris-petrov** [comment]
Ah, yes, sorry, I read the message and because `public` *is* necessary, I thought it a false-positive. So yes, indeed, only the message is wrong.

**Comment by adangel** [comment]
That's correct.
There's one more thing you can do: enhancing the existing unit test to verify the message, too. That way, we would have spotted the problem earlier.
Simply add an additional test case in https://github.com/pmd/pmd/blob/master/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryModifier.xml at the end. Look at the last test case, to see, how the messages can be verified via `expected-messages`.
Thanks for tackling this!

**Comment by vishva007** [comment]
Hi,

This is the test case I am adding in the xml file 

<test-code>
        <description>Static Modifier on Nested Interface</description>
        <expected-problems>1</expected-problems>
        <expected-messages>
            <message>Unnecessary modifier 'static' on interface 'Bar': types nested within an interface type are implicitly static</message>
        </expected-messages>
        <code><![CDATA[
class Foo{
    public static interface Bar{
        void method(){ }
    }
}
        ]]></code>
    </test-code>

Is this good ?

Thank you

**Comment by adangel** [comment]
Looks good!

**Comment by ToryZhou** [comment]
Remove the 'public' on the interface. it worked.
